crypto/cipher_extra/e_aes_cbc_hmac_sha1.c

```diff
@@ -0,0 +1,375 @@
+  int ret;
+
+  int key_bits = EVP_CIPHER_CTX_key_length(ctx) * 8;
+  if (enc) {
+    ret = aes_hw_set_encrypt_key(inkey, key_bits, &key->ks);
+  } else {
+    ret = aes_hw_set_decrypt_key(inkey, key_bits, &key->ks);
+  }
+  if (ret < 0) {
+    return 0;
+  }
+
+  SHA1_Init(&key->head);
+  key->tail = key->head;
+  key->md = key->head;
+
+  key->payload_length = NO_PAYLOAD_LENGTH;
+
+  return 1;
+}
+
+static int aesni_cbc_hmac_sha1_cipher(EVP_CIPHER_CTX *ctx, unsigned char *out,
+                                      const unsigned char *in, size_t len) {
+  EVP_AES_HMAC_SHA1 *key = (EVP_AES_HMAC_SHA1 *)(ctx->cipher_data);
+  size_t plen = key->payload_length, iv = 0;
+
+  key->payload_length = NO_PAYLOAD_LENGTH;
+
+  if (len % AES_BLOCK_SIZE) {
+    OPENSSL_PUT_ERROR(CIPHER, CIPHER_R_DATA_NOT_MULTIPLE_OF_BLOCK_LENGTH);
+    return 0;
+  }
+
+  if (EVP_CIPHER_CTX_encrypting(ctx)) {
+    // NOTE: Difference between openssl and aws-lc:
+    // In encrypt case, |plen| is set in the call |EVP_CIPHER_CTX_ctrl| with
+    // |EVP_CTRL_AEAD_TLS1_AAD| operation.
+    // When |plen == NO_PAYLOAD_LENGTH|, it means the call did not happen.
+    // In this case, aws-lc returns error(0) but openssl supports that with
+    // below explanation.
+    // https://mta.openssl.org/pipermail/openssl-users/2019-November/011458.html
+    // -- These stitched ciphers are specifically targeted at use by libssl
+    //    and are designed for use in SSL/TLS only.
+    if (plen == NO_PAYLOAD_LENGTH) {
+      // |EVP_CIPHER_CTX_ctrl| with |EVP_CTRL_AEAD_TLS1_AAD| operation is not
+      // performed.
+      OPENSSL_PUT_ERROR(CIPHER, CIPHER_R_CTRL_OPERATION_NOT_PERFORMED);
+      return 0;
+    }
+    if (len !=
+        ((plen + SHA_DIGEST_LENGTH + AES_BLOCK_SIZE) & -AES_BLOCK_SIZE)) {
+      // The input should have space for plen(eiv + plaintext) + SHA_DIGEST_LENGTH + padding.
+      OPENSSL_PUT_ERROR(CIPHER, CIPHER_R_UNSUPPORTED_INPUT_SIZE);
+      return 0;
+    } else if (key->aux.tls_ver >= TLS1_1_VERSION) {
+      iv = AES_BLOCK_SIZE;
+    }
+
+    size_t aes_off = 0;
+    size_t sha_off = SHA_CBLOCK - key->md.num;
+    size_t blocks;
+
+    // Use stitch code |aesni_cbc_sha1_enc| when there are multiple of SHA_CBLOCK
+    // so |aesni_cbc_sha1_enc| can use AES and SHA on the same data block.
+    //
+    // Assembly stitch handles AVX-capable processors, but its
+    // performance is not optimal on AMD Jaguar, ~40% worse, for
+    // unknown reasons. Incidentally processor in question supports
+    // AVX, but not AMD-specific XOP extension, which can be used
+    // to identify it and avoid stitch invocation. So that after we
+    // establish that current CPU supports AVX, we even see if it's
+    // either even XOP-capable Bulldozer-based or GenuineIntel one.
+    // But SHAEXT-capable go ahead...
+    if ((CRYPTO_is_SHAEXT_capable() ||
+        (CRYPTO_is_AVX_capable() &&
+        (CRYPTO_is_AMD_XOP_support() | CRYPTO_is_intel_cpu()))) &&
```

Comment:
```
    // Assembly stitch handles AVX-capable processors, but its
    // performance is not optimal on AMD Jaguar, ~40% worse, for
    // unknown reasons. Incidentally processor in question supports
    // AVX, but not AMD-specific XOP extension, which can be used
    // to identify it and avoid stitch invocation. So that after we
    // establish that current CPU supports AVX, we even see if it's
    // either even XOP-capable Bulldozer-based or GenuineIntel one.
    // But SHAEXT-capable go ahead...
```

Reply:
fyi, XOP bit is always disabled: https://github.com/awslabs/aws-lc/blob/506ed23574a6323bd97f406416603f548cb8ba56/crypto/fipsmodule/cpucap/cpu_intel.c#L232.

Reply:
Discussed with Torben, we are good to keep bitwise instead of using or.